File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int answer;

    int maxPathSum(TreeNode* root) {
        answer = INT_MIN;
        dfs(root);
        return answer;
    }

    int dfs(TreeNode* root) {
        if (root == NULL) return 0;
        int left = (root->left == NULL) ? 0 : dfs(root->left);
        int right = (root->right == NULL) ? 0 : dfs(root->right);
        int maxPathSumAtNode = max({root->val, root->val + left, root->val + right, root->val + left + right});
        answer = max(answer, maxPathSumAtNode);
        return max({root->val, root->val + left, root->val + right, 0});
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
class Solution {
    public int maxPathSum(TreeNode root) {
        int[] pathSum = new int[] {Integer.MIN_VALUE};
        maxPathSum(root, pathSum);
        return pathSum[0];
    }

    private int maxPathSum(TreeNode root, int[] pathSum) {
        if (root == null) {
            return Integer.MIN_VALUE;
        }
        int left = maxPathSum(root.left, pathSum);
        int right = maxPathSum(root.right, pathSum);
        int leftPathSum = (left == Integer.MIN_VALUE ? 0 : left);
        int rightPathSum = (right == Integer.MIN_VALUE ? 0 : right);
        int leftChildIncluded = leftPathSum + root.val;
        int rightChildIncluded = rightPathSum + root.val;
        int completePath = leftPathSum + rightPathSum + root.val;
        int maxPathSum = Math.max(Math.max(leftChildIncluded, rightChildIncluded), Math.max(completePath, root.val));
        pathSum[0] = Math.max(pathSum[0], maxPathSum);
        int returnToTop = Math.max(root.val, Math.max(leftChildIncluded, rightChildIncluded));
        // System.out.println("Node: " + root.val + " send: " + returnToTop + " maxTillNow: " + pathSum[0]);
        return returnToTop;
    }
}